walisc/JeMPI
# Dev delete indexs

**Ticket:** N/A
**Other Related Tickets:** N/A

----

## Describe of changes

This change adds to the ability to modify  indexes (in dgraph) before batch linking occurs. This allows you to remove unnecessary indexes hence speeding up the linking processes (as indexes do not have to be updated per insert/update). Once this linking process in complete it will revert back to the original indexes. 

Items to note: 
- for updating indexes before linking, your linking rules in your config should not contain any  probabilistic rules. If this is set, the default indexes will be used. 
- In addition you will need to add an indexes property to you config specifying the fields you want to index whilst linking. An example of this is below

```json
"indexes": {
    "linking": {
      "national_id": {
        "props": "@index(exact,trigram)"
      }
    }
  },
```
## How to test

- Update you config to ensure that your linking rules contain not probabilistic rules
- Run the configurator in `JeMPI_Apps/JeMPI_Configuration` i,e `./create.sh reference/config-reference.json`
- Start JeMPI
- Upload a csv or add a csv to the async csv folder
- Wait for linking to begin, and then check the linker logs. You should see a log similar to the following
` 2024-03-06 03:10:28.786 LibMPIIndexesHooks:20 - Updating indexes for linking.`

- using ratel (ui for dgraph), look at the schema whilst its linking. Only the nationId (and SourceId.facility and SourceId.patient) should have indexes

![image](https://github.com/walisc/JeMPI/assets/5619335/ecd4dc19-1c49-49de-97b7-f7d7edb48aaf)


- Wait for linking to complete. You should see a log similar to the following
 `LibMPIIndexesHooks:29 - Adding default indexes after linking.`

- using ratel (ui for dgraph), look at the schema it should have reverted back to the default
![image](https://github.com/walisc/JeMPI/assets/5619335/c3e7fafc-e906-414b-9500-9dd7ffe0cca7)

**Please Note:** There is an issue with the CustomeMU disabling the linker when an error occurs. This will also need to be address when testing. Currently set `SEND_INTERACTIONS_TO_EM` to `false`
